arbos/addressSet/addressSet_test.go

```diff
@@ -173,6 +174,135 @@ func TestAddressSetAllMembers(t *testing.T) {
+	}
+	// Run RectifyMapping to fix the issue
+	checkIfRectifyMappingWorks(t, addr3, aset, 1, 1)
+	Require(t, aset.Clear())
+
+	// Test Goerli history
+	addr1 = common.HexToAddress("0x186B56023d42B2B4E7616589a5C62EEf5FCa21DD")
+	addr2 = common.HexToAddress("0xc8efdb677afeb775ce1617dd976b56b3a6e95bba")
+	addr3 = common.HexToAddress("0xc3f86bb81e32295d29c288ffb4828936538cf326")
+	addr4 = common.HexToAddress("0x67acb531a05160a81dcd03079347f264c4fa2da3")
+	// Follow logs
+	Require(t, aset.Add(addr1))
+	Require(t, aset.Add(addr2))
+	Require(t, aset.Add(addr3))
+	Require(t, aset.Remove(addr1, version))
+	Require(t, aset.Add(addr4))
+	Require(t, aset.Remove(addr3, version))
+	Require(t, aset.Remove(addr2, version))
+	// Check if history's correct
+	CurrentOwner, _ = aset.backingStorage.GetByUint64(uint64(1))
+	isOwner, _ = aset.IsMember(addr3)
+	correctOwner, _ = aset.IsMember(addr4)
+	if size(t, aset) != uint64(1) || CurrentOwner != common.BytesToHash(addr3.Bytes()) || isOwner || !correctOwner {
+		Fail(t, "Logs and current state did not match")
+	}
+	// Run RectifyMapping to fix the issue
+	checkIfRectifyMappingWorks(t, addr4, aset, 1, 1)
+}
+
+func TestRectifyMapping(t *testing.T) {
+	db := storage.NewMemoryBackedStateDB()
+	sto := storage.NewGeth(db, burn.NewSystemBurner(nil, false))
+	Require(t, Initialize(sto))
+	aset := OpenAddressSet(sto)
+
+	addr1 := testhelpers.RandomAddress()
+	addr2 := testhelpers.RandomAddress()
+	addr3 := testhelpers.RandomAddress()
+	possibleAddresses := []common.Address{addr1, addr2, addr3}
+
+	Require(t, aset.Add(addr1))
+	Require(t, aset.Add(addr2))
+	Require(t, aset.Add(addr3))
+
+	// RectifyMapping should not do anything if the mapping is correct
+	addr := possibleAddresses[rand.Intn(len(possibleAddresses))]
+	Require(t, aset.RectifyMapping(addr))
+
+	// Non owner's should not be able to call RectifyMapping
+	err := aset.RectifyMapping(testhelpers.RandomAddress())
+	if err == nil {
+		Fail(t, "RectifyMapping was succesfully called by non owner")
+	}
+
+	// Corrupt the list and verify if RectifyMapping fixes it
+	addrHash := common.BytesToHash(addr2.Bytes())
+	Require(t, aset.backingStorage.SetByUint64(uint64(1), addrHash))
+	checkIfRectifyMappingWorks(t, addr1, aset, 3, 1)
+
+	// Corrupt the map and verify if RectifyMapping fixes it
+	addrHash = common.BytesToHash(addr2.Bytes())
+	Require(t, aset.byAddress.Set(addrHash, util.UintToHash(uint64(6))))
+	checkIfRectifyMappingWorks(t, addr2, aset, 3, 2)
+
+	// Add a new owner to the map and verify if RectifyMapping syncs list with the map
+	// to check for the case where list has fewer owners than expected
+	addr4 := testhelpers.RandomAddress()
+	addrHash = common.BytesToHash(addr4.Bytes())
+	Require(t, aset.byAddress.Set(addrHash, util.UintToHash(uint64(1))))
+	checkIfRectifyMappingWorks(t, addr4, aset, 4, 4)
+}
+
+func checkIfRectifyMappingWorks(t *testing.T, addr common.Address, aset *AddressSet, asetSize uint64, ownerIndex uint64) {
```

Comment:
I'm not sure where we can put this check, but could we add a check to make sure that AllMembers() returns the correct list after RectifyMapping? Especially for the real history tests

Reply:
I've added the test inside `checkIfRectifyMappingWorks`